File: app/src/main/java/uk/ac/swansea/codenames/join_game.java
package uk.ac.swansea.codenames;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.ShapeDrawable;
import android.os.Bundle;
import android.os.Handler;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Space;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.res.ResourcesCompat;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;

import io.socket.emitter.Emitter;

public class join_game extends AppCompatActivity {
    private ConstraintLayout constraintLayout;
    private LinearLayout scrollLinear;
    private Button backButton;
    private Button refreshButton;
    private Button joinPrivateButton;
    private TextView joinGameTitle;
    private TextView publicRoomText;
    private TextView privateRoomText;
    private EditText roomNameEdit;
    private EditText passwordEdit;

    private ArrayList<Button> publicJoinButtons = new ArrayList<>();

    private int defaultColour;
    private String username;
    private boolean validJoin;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.join_game);

        constraintLayout = findViewById(R.id.constraintLayout);
        scrollLinear = findViewById(R.id.scrollLinear);
        backButton = findViewById(R.id.backButton);
        refreshButton = findViewById(R.id.refreshButton);
        joinGameTitle = findViewById(R.id.joinGameTitle);
        joinPrivateButton = findViewById(R.id.joinPrivateButton);
        publicRoomText = findViewById(R.id.publicRoomText);
        privateRoomText = findViewById(R.id.privateRoomText);
        roomNameEdit = findViewById(R.id.roomNameEdit);
        passwordEdit = findViewById(R.id.passwordEdit);

        username = userSettings.getInstance().getPreference("username");

        socketConnection.socket.emit("getAllRooms");

        joinPrivateButton.setOnClickListener(v -> {
            validJoin = true;

            socketConnection.socket.emit("joinRoom", username, roomNameEdit.getText().toString(),
                    passwordEdit.getText().toString());

            backButton.setEnabled(false);
            refreshButton.setEnabled(false);
            roomNameEdit.setEnabled(false);
            passwordEdit.setEnabled(false);
            joinPrivateButton.setEnabled(false);

            for (Button b : publicJoinButtons) {
                b.setEnabled(false);
            }

            Handler handler = new Handler();

            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    if (validJoin) {
                        Intent i = new Intent(getApplicationContext(), online_game.class);
                        i.putExtra("username", userSettings.getInstance().getPreference("username"));
                        i.putExtra("roomName", roomNameEdit.getText().toString());
                        i.putExtra("isHost", false);
                        startActivity(i);
                    } else {
                        backButton.setEnabled(true);
                        refreshButton.setEnabled(true);
                        roomNameEdit.setEnabled(true);
                        passwordEdit.setEnabled(true);
                        joinPrivateButton.setEnabled(true);
                    }
                }
            }, 3000);
        });

        refreshButton.setOnClickListener(v -> {
            scrollLinear.removeAllViews();

            socketConnection.socket.emit("getAllRooms");
        });

        socketConnection.socket.on("joinFail", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                toggleMessageBox((String) args[0]);
                validJoin = false;
            }
        });

        socketConnection.socket.on("joinFailNick", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                toggleNicknameBox((String) args[0]);
                validJoin = false;
            }
        });

        socketConnection.socket.on("allRooms", new Emitter.Listener() {
            @Override
            public void call(Object... args) {
                publicJoinButtons.clear();

                JSONObject data = (JSONObject) args[0];

                ArrayList<String> allRoomNames = new ArrayList<>();
                ArrayList<String> publicRooms = new ArrayList<>();

                JSONArray jArray = (JSONArray) data.names();

                if (jArray != null) {
                    for (int i = 0; i < jArray.length(); i++) {
                        try {
                            allRoomNames.add(jArray.getString(i));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }

                for (String s : allRoomNames) {
                    try {
                        JSONObject j = (JSONObject) data.get(s);

                        String temp = j.toString();

                        String[] parts = temp.split(",");

                        parts[0] = parts[0].replace("{\"roomName\":", "");
                        parts[1] = parts[1].replace("\"password\":", "");
                        parts[3] = parts[3].replace("\"closed\":", "");
                        parts[3] = parts[3].replace("}", "");

                        String roomName = parts[0].replace("\"", "");
                        String password = parts[1].replace("\"", "");
                        boolean closed = Boolean.parseBoolean(parts[3]);

                        if (password.equals("") && !closed) {
                            publicRooms.add(roomName);
                        }

                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        for (String name : publicRooms) {
                            LinearLayout roomLinear = new LinearLayout(getApplicationContext());
                            roomLinear.setOrientation(LinearLayout.HORIZONTAL);

                            if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT).equals("")) {
                                defaultColour = userSettings.getInstance().MENU_TEXT_DEFAULT;
                            } else {
                                defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT));
                            }

                            Button joinButton = new Button(getApplicationContext());
                            joinButton.setText(name);
                            joinButton.setTextColor(defaultColour);

                            LinearLayout.LayoutParams buttonParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);

                            joinButton.setLayoutParams(buttonParams);

                            if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS).equals("")) {
                                defaultColour = userSettings.getInstance().MENU_BUTTONS_DEFAULT;
                            } else {
                                defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS));
                            }

                            joinButton.setBackgroundColor(defaultColour);

                            joinButton.setOnClickListener(v -> {
                                validJoin = true;

                                socketConnection.socket.emit("joinRoom", username, name, "");

                                backButton.setEnabled(false);
                                refreshButton.setEnabled(false);
                                roomNameEdit.setEnabled(false);
                                passwordEdit.setEnabled(false);
                                joinPrivateButton.setEnabled(false);

                                for (Button b : publicJoinButtons) {
                                    b.setEnabled(false);
                                }

                                Handler handler = new Handler();

                                handler.postDelayed(new Runnable() {
                                    @Override
                                    public void run() {
                                        if (validJoin) {
                                            Intent i = new Intent(getApplicationContext(), online_game.class);
                                            i.putExtra("username", userSettings.getInstance().getPreference("username"));
                                            i.putExtra("roomName", name);
                                            i.putExtra("isHost", false);
                                            startActivity(i);
                                        } else {
                                            backButton.setEnabled(true);
                                            refreshButton.setEnabled(true);
                                            roomNameEdit.setEnabled(true);
                                            passwordEdit.setEnabled(true);
                                            joinPrivateButton.setEnabled(true);
                                        }
                                    }
                                }, 3000);
                            });

                            publicJoinButtons.add(joinButton);

                            scrollLinear.addView(roomLinear);
                            roomLinear.addView(joinButton);
                        }
                    }
                });
            }
        });

        updateColours();
    }

    @Override
    public void onBackPressed() {
        backButton(getWindow().getDecorView());
    }

    public void backButton(View view) {
        Intent i = new Intent(view.getContext(), online_setup.class);
        startActivity(i);
    }

    private void toggleMessageBox(String message) {

    }

    private void toggleNicknameBox(String message) {

    }

    private void updateColours() {
        if (userSettings.getInstance().getPreference(userSettings.getInstance().APPLICATION_BACKGROUND).equals("")) {
            defaultColour = userSettings.getInstance().APPLICATION_BACKGROUND_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().APPLICATION_BACKGROUND));
        }

        constraintLayout.setBackgroundColor(defaultColour);

        if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS).equals("")) {
            defaultColour = userSettings.getInstance().MENU_BUTTONS_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_BUTTONS));
        }

        backButton.setBackgroundColor(defaultColour);
        refreshButton.setBackgroundColor(defaultColour);
        joinPrivateButton.setBackgroundColor(defaultColour);

        if (userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT).equals("")) {
            defaultColour = userSettings.getInstance().MENU_TEXT_DEFAULT;
        } else {
            defaultColour = Integer.parseInt(userSettings.getInstance().getPreference(userSettings.getInstance().MENU_TEXT));
        }

        joinGameTitle.setTextColor(defaultColour);
        publicRoomText.setTextColor(defaultColour);
        privateRoomText.setTextColor(defaultColour);
        roomNameEdit.setTextColor(defaultColour);
        passwordEdit.setTextColor(defaultColour);
        backButton.setTextColor(defaultColour);
        refreshButton.setTextColor(defaultColour);
        joinPrivateButton.setTextColor(defaultColour);
    }
}

File: app/src/main/java/uk/ac/swansea/codenames/JoinGame.kt
package uk.ac.swansea.codenames

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import android.widget.LinearLayout
import android.os.Bundle
import android.content.Intent
import android.media.MediaPlayer
import android.os.Handler
import android.os.Looper
import android.speech.tts.TextToSpeech
import android.view.View
import org.json.JSONObject
import org.json.JSONArray
import org.json.JSONException
import android.widget.Button
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textview.MaterialTextView
import java.util.*

/**
 * Scene for joining online games.
 */
class JoinGame : AppCompatActivity(), TextToSpeech.OnInitListener {
    private var constraintLayout: ConstraintLayout? = null
    private var messageBox: ConstraintLayout? = null
    private var scrollLinear: LinearLayout? = null
    private var backButton: MaterialButton? = null
    private var refreshButton: MaterialButton? = null
    private var joinPrivateButton: MaterialButton? = null
    private var okButton: MaterialButton? = null
    private var joinGameTitle: MaterialTextView? = null
    private var publicRoomText: MaterialTextView? = null
    private var privateRoomText: MaterialTextView? = null
    private var messageText: MaterialTextView? = null
    private var nicknameEdit: TextInputEditText? = null
    private var roomNameEdit: TextInputEditText? = null
    private var passwordEdit: TextInputEditText? = null
    private val publicJoinButtons = ArrayList<Button>()
    private var applicationBackgroundColour = -10921639
    private var menuButtonsColour = -8164501
    private var menuTextColour = -1
    private var username: String? = null
    private var validJoin = false
    private var messageBoxOpen = false
    private var textToSpeech: TextToSpeech? = null
    private var buttonClick: MediaPlayer? = null

    /**
     * Sets up layout and listeners etc.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.join_game)

        constraintLayout = findViewById(R.id.constraintLayout)
        messageBox = findViewById(R.id.messageBox)
        scrollLinear = findViewById(R.id.scrollLinear)
        backButton = findViewById(R.id.backButton)
        refreshButton = findViewById(R.id.refreshButton)
        joinGameTitle = findViewById(R.id.joinGameTitle)
        joinPrivateButton = findViewById(R.id.joinPrivateButton)
        publicRoomText = findViewById(R.id.publicRoomText)
        privateRoomText = findViewById(R.id.privateRoomText)
        nicknameEdit = findViewById(R.id.nicknameEdit)
        roomNameEdit = findViewById(R.id.roomNameEdit)
        passwordEdit = findViewById(R.id.passwordEdit)
        okButton = findViewById(R.id.okButton)
        messageText = findViewById(R.id.messageText)

        buttonClick = MediaPlayer.create(this, R.raw.buttonclick)

        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        val soundFXVolume = preferences.getFloat("soundFXVolume", 0.5f)

        buttonClick?.setVolume(soundFXVolume, soundFXVolume)

        textToSpeech = TextToSpeech(this, this)

        username = preferences.getString("username", "")

        nicknameEdit?.setText(username)

        SocketConnection.socket.emit("getAllRooms")

        joinPrivateButton?.setOnClickListener {
            buttonClick?.start()

            validJoin = true

            SocketConnection.socket.close()
            SocketConnection.socket.open()

            SocketConnection.socket.emit("joinRoom", username, roomNameEdit?.text.toString(),
                    passwordEdit?.text.toString())

            backButton?.isEnabled = false
            refreshButton?.isEnabled = false
            roomNameEdit?.isEnabled = false
            passwordEdit?.isEnabled = false
            joinPrivateButton?.isEnabled = false

            for (b in publicJoinButtons) {
                b.isEnabled = false
            }

            Handler(Looper.getMainLooper()).postDelayed({
                if (validJoin) {
                    textToSpeech?.stop()

                    val i = Intent(applicationContext, OnlineGame::class.java)
                    i.putExtra("username", username)
                    i.putExtra("roomName", roomNameEdit?.text.toString())
                    i.putExtra("isHost", false)

                    startActivity(i)
                } else {
                    backButton?.isEnabled = true
                    refreshButton?.isEnabled = true
                    roomNameEdit?.isEnabled = true
                    passwordEdit?.isEnabled = true
                    joinPrivateButton?.isEnabled = true
                }
            }, 3000)
        }

        refreshButton?.setOnClickListener {
            buttonClick?.start()

            scrollLinear?.removeAllViews()
            SocketConnection.socket.emit("getAllRooms")
        }

        backButton?.setOnClickListener {
            buttonClick?.start()

            if (!messageBoxOpen) {
                textToSpeech?.stop()

                val i = Intent(applicationContext, OnlineSetup::class.java)
                startActivity(i)
            }
        }

        okButton?.setOnClickListener {
            buttonClick?.start()

            textToSpeech?.stop()

            messageBoxOpen = false

            messageBox?.visibility = View.INVISIBLE

            for (b in publicJoinButtons) {
                b.isEnabled = true
            }

            val editor = preferences.edit()

            editor.putString("username", nicknameEdit?.text.toString())

            editor.apply()

            username = preferences.getString("username", "")
        }

        SocketConnection.socket.on("joinFail") { args ->
            toggleMessageBox(args[0] as String)
            validJoin = false
        }

        SocketConnection.socket.on("joinFailNick") { args ->
            toggleNicknameBox(args[0] as String)
            validJoin = false
        }

        SocketConnection.socket.on("allRooms") { args ->
            publicJoinButtons.clear()

            val data = args[0] as JSONObject

            if (data.length() > 0) {
                val allRoomNames = ArrayList<String>()
                val publicRooms = ArrayList<String>()
                val jArray = data.names() as JSONArray

                for (i in 0 until jArray.length()) {
                    try {
                        allRoomNames.add(jArray.getString(i))
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }

                for (s in allRoomNames) {
                    try {
                        val j = data[s] as JSONObject
                        val temp = j.toString()
                        val parts = temp.split(",").toTypedArray()

                        parts[0] = parts[0].replace("{\"roomName\":", "")
                        parts[1] = parts[1].replace("\"password\":", "")
                        parts[3] = parts[3].replace("\"closed\":", "")
                        parts[3] = parts[3].replace("}", "")

                        val roomName = parts[0].replace("\"", "")
                        val password = parts[1].replace("\"", "")
                        val closed = parts[3].toBoolean()

                        if (password == "" && !closed) {
                            publicRooms.add(roomName)
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }

                //Add buttons for every public game found.
                runOnUiThread {
                    for (name in publicRooms) {
                        val roomLinear = LinearLayout(applicationContext)
                        roomLinear.orientation = LinearLayout.HORIZONTAL

                        menuButtonsColour = preferences.getInt("menuButton", -8164501)
                        menuTextColour = preferences.getInt("menuText", -1)

                        val joinButton = Button(applicationContext)
                        joinButton.text = name
                        joinButton.setTextColor(menuTextColour)
                        joinButton.setBackgroundColor(menuButtonsColour)
                        joinButton.isSoundEffectsEnabled = false

                        val buttonParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)

                        buttonParams.bottomMargin = 8

                        joinButton.layoutParams = buttonParams

                        joinButton.setOnClickListener {
                            buttonClick?.start()

                            validJoin = true

                            SocketConnection.socket.close()
                            SocketConnection.socket.open()

                            SocketConnection.socket.emit("joinRoom", username, name, "")

                            backButton?.isEnabled = false
                            refreshButton?.isEnabled = false
                            roomNameEdit?.isEnabled = false
                            passwordEdit?.isEnabled = false
                            joinPrivateButton?.isEnabled = false

                            for (b in publicJoinButtons) {
                                b.isEnabled = false
                            }

                            Handler(Looper.getMainLooper()).postDelayed({
                                if (validJoin) {
                                    textToSpeech?.stop()

                                    val i = Intent(applicationContext, OnlineGame::class.java)
                                    i.putExtra("username", username)
                                    i.putExtra("roomName", name)
                                    i.putExtra("isHost", false)

                                    startActivity(i)
                                } else {
                                    backButton?.isEnabled = true
                                    refreshButton?.isEnabled = true
                                    roomNameEdit?.isEnabled = true
                                    passwordEdit?.isEnabled = true
                                    joinPrivateButton?.isEnabled = true
                                }
                            }, 3000)
                        }

                        joinButton.setOnLongClickListener {
                            speakOut(joinButton.text.toString())
                            true
                        }

                        publicJoinButtons.add(joinButton)
                        scrollLinear?.addView(roomLinear)
                        roomLinear.addView(joinButton)
                    }
                }
            }
        }

        backButton?.setOnLongClickListener {
            speakOut(backButton?.text.toString())
            true
        }

        joinGameTitle?.setOnLongClickListener {
            speakOut(joinGameTitle?.text.toString())
            true
        }

        refreshButton?.setOnLongClickListener {
            speakOut(refreshButton?.text.toString())
            true
        }

        publicRoomText?.setOnLongClickListener {
            speakOut(publicRoomText?.text.toString())
            true
        }

        privateRoomText?.setOnLongClickListener {
            speakOut(privateRoomText?.text.toString())
            true
        }

        joinPrivateButton?.setOnLongClickListener {
            speakOut(joinPrivateButton?.text.toString())
            true
        }

        messageText?.setOnLongClickListener {
            speakOut(messageText?.text.toString())
            true
        }

        okButton?.setOnLongClickListener {
            speakOut(okButton?.text.toString())
            true
        }

        updateColours()
    }

    /**
     * When app is reopened after being minimised.
     */
    override fun onResume() {
        super.onResume()

        val intent = Intent(this, BackgroundMusicService::class.java)

        intent.action = "RESUME"

        startService(intent)
    }

    /**
     * When app is minimised.
     */
    override fun onPause() {
        super.onPause()

        val intent = Intent(this, BackgroundMusicService::class.java)

        intent.action = "PAUSE"

        startService(intent)
    }

    /**
     * Sets up Text-to-Speech engine.
     */
    override fun onInit(status: Int) {
        if (status == TextToSpeech.SUCCESS) {
            textToSpeech!!.language = Locale.UK
        }
    }

    /**
     * When device's back button pressed.
     */
    override fun onBackPressed() {
        backButton?.performClick()
    }

    /**
     * Toggles the message box being opened.
     */
    private fun toggleMessageBox(message: String) {
        messageBoxOpen = true

        runOnUiThread {
            speakOut(messageText?.text.toString())

            messageBox?.visibility = View.VISIBLE
            messageText?.text = message
            nicknameEdit?.visibility = View.GONE

            for (b in publicJoinButtons) {
                b.isEnabled = false
            }
        }
    }

    /**
     * Toggles the nickname box being opened.
     */
    private fun toggleNicknameBox(message: String) {
        messageBoxOpen = true

        runOnUiThread {
            speakOut(messageText?.text.toString())

            messageBox?.visibility = View.VISIBLE
            messageText?.text = message
            nicknameEdit?.visibility = View.VISIBLE

            for (b in publicJoinButtons) {
                b.isEnabled = false
            }
        }
    }

    /**
     * Reads aloud the given text.
     */
    private fun speakOut(message: String) {
        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        if (preferences.getBoolean("textToSpeech", true)) {
            textToSpeech!!.speak(message, TextToSpeech.QUEUE_FLUSH, null, "")
        }
    }

    /**
     * Updates colours of elements in the layout.
     */
    private fun updateColours() {
        val preferences = this.getSharedPreferences("preferences", Context.MODE_PRIVATE)

        applicationBackgroundColour = preferences.getInt("applicationBackground", -10921639)
        menuButtonsColour = preferences.getInt("menuButton", -8164501)
        menuTextColour = preferences.getInt("menuText", -1)

        constraintLayout?.setBackgroundColor(applicationBackgroundColour)
        messageBox?.setBackgroundColor(applicationBackgroundColour)

        backButton?.setBackgroundColor(menuButtonsColour)
        refreshButton?.setBackgroundColor(menuButtonsColour)
        joinPrivateButton?.setBackgroundColor(menuButtonsColour)
        okButton?.setBackgroundColor(menuButtonsColour)

        joinGameTitle?.setTextColor(menuTextColour)
        publicRoomText?.setTextColor(menuTextColour)
        privateRoomText?.setTextColor(menuTextColour)
        messageText?.setTextColor(menuTextColour)
        roomNameEdit?.setTextColor(menuTextColour)
        passwordEdit?.setTextColor(menuTextColour)
        nicknameEdit?.setTextColor(menuTextColour)
        backButton?.setTextColor(menuTextColour)
        refreshButton?.setTextColor(menuTextColour)
        joinPrivateButton?.setTextColor(menuTextColour)
        okButton?.setTextColor(menuTextColour)
    }
}